File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys
from collections import deque

input = sys.stdin.readline
matrix = []
visited = []
N = 0
M = 0
vector = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def bfs(row, col):
    q = deque()
    q.append([row, col])
    visited[row][col] = 1
    while q:
        now_row, now_col = q.popleft()
        for dr, dc in vector:
            next_row = now_row + dr
            next_col = now_col + dc
            if -1 < next_row < N and -1 < next_col < M and matrix[next_row][next_col] and not visited[next_row][
                next_col]:
                q.append([next_row, next_col])
                visited[next_row][next_col] = 1


def solution():
    global matrix, visited, N, M
    M, N, K = map(int, input().split())
    matrix = [[0] * M for _ in range(N)]
    for _ in range(K):
        X, Y = map(int, input().split())
        matrix[Y][X] = 1
    visited = [[0] * M for _ in range(N)]
    count = 0

    for row in range(N):
        for col in range(M):
            if matrix[row][col] and not visited[row][col]:
                bfs(row, col)
                count += 1

    return count


for _ in range(int(input())):
    print(solution())

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf8').trim().split('\n');

const T = +input[0];
let j = 1;

function solution(M, N, K, cabbages) {
  let result = 0;
  const matrix = new Array(N).fill(0).map(() => new Array(M).fill(0));
  const visited = new Array(N).fill(0).map(() => new Array(M).fill(0));
  cabbages.forEach((cabbage) => {
    const [X, Y] = cabbage;
    matrix[Y][X] = 1;
  });
  const vector = [
    [-1, 0],
    [1, 0],
    [0, -1],
    [0, 1],
  ];

  function bfs(startRow, startCol) {
    const queue = [[startRow, startCol]];
    let idx = 0;
    visited[startRow][startCol] = 1;

    while (idx < queue.length) {
      const [row, col] = queue[idx++];
      vector.forEach((v) => {
        const [dr, dc] = v;
        const nextRow = row + dr;
        const nextCol = col + dc;
        if (
          -1 < nextRow &&
          nextRow < N &&
          -1 < nextCol &&
          nextCol < M &&
          matrix[nextRow][nextCol] &&
          !visited[nextRow][nextCol]
        ) {
          queue.push([nextRow, nextCol]);
          visited[nextRow][nextCol] = 1;
        }
      });
    }
  }

  for (let row = 0; row < N; row++) {
    for (let col = 0; col < M; col++) {
      if (!visited[row][col] && matrix[row][col]) {
        bfs(row, col);
        result++;
      }
    }
  }
  return result;
}

for (let i = 0; i < T; i++) {
  const [M, N, K] = input[j].split(' ').map((v) => +v);
  const cabbages = input
    .slice(j + 1, j + K + 1)
    .map((cabbage) => cabbage.split(' ').map((v) => +v));
  console.log(solution(M, N, K, cabbages));
  j += K + 1;
}
